plugins/spi/SPIOutput.cpp

```diff
@@ -486,6 +500,100 @@ uint8_t SPIOutput::P9813CreateFlag(uint8_t red, uint8_t green, uint8_t blue) {
   return ~flag;
 }
 
+
+void SPIOutput::IndividualAPA102Control(const DmxBuffer &buffer) {
+  // some detailed information on the protocol:
+  // https://cpldcpu.wordpress.com/2014/11/30/understanding-the-apa102-superled/
+  // Data-Struct
+  // StartFrame: 4Byte = 32Bits zeros
+  // LEDFrame: 1Byte FF ; 3Byte color info (Blue, Green, Red)
+  // EndFrame: (n/2)bits; n = pixel_count
+  
+  // i don't know how this works.. but it works ;-)
+  const uint8_t latch_bytes = 3 * APA102_SPI_BYTES_PER_PIXEL;
+  const unsigned int first_slot = m_start_address - 1;  // 0 offset
+  
+  // only do something if minimum 1pixel can be updated..
+  if (buffer.Size() - first_slot < APA102_SLOTS_PER_PIXEL) {
+    // not even 3 bytes of data, don't bother updating
+    return;
+  }
+  
+  // We always check out the entire string length, even if we only have data
+  // for part of it
+  const unsigned int output_length = m_pixel_count * APA102_SPI_BYTES_PER_PIXEL;
+  uint8_t *output = m_backend->Checkout(m_output_number, output_length,
+                                        latch_bytes);
+  // only update spi data if possible
+  if (!output)
+    return;
+
+  for (unsigned int i = 0; i < m_pixel_count; i++) {
+    // Convert RGB to APA102 Pixel
+    unsigned int offset = first_slot + i * APA102_SLOTS_PER_PIXEL;
+    // We need to avoid the first 4 bytes of the buffer since that acts as a
+    // start of frame delimiter
+    unsigned int spi_offset = (i + 1) * APA102_SPI_BYTES_PER_PIXEL;
+    uint8_t r = 0;
+    uint8_t b = 0;
+    uint8_t g = 0;
+    if (buffer.Size() - offset >= APA102_SLOTS_PER_PIXEL) {
+      r = buffer.Get(offset);
+      g = buffer.Get(offset + 1);
+      b = buffer.Get(offset + 2);
+    }
+    // first Byte contains:
+    // 3bit start mark (111) + 5bit GlobalBrightnes
+    // set GlobalBrightnes fixed to 31 --> that reduces flickering
+	// that can be written as 0xE0 & 0x1F
+    output[spi_offset + 0] = 0xFF;
+    output[spi_offset + 1] = b;
+    output[spi_offset + 2] = g;
+    output[spi_offset + 3] = r;
+  }
+  m_backend->Commit(m_output_number);
+}
+
+void SPIOutput::CombinedAPA102Control(const DmxBuffer &buffer) {
+  // for Protocol details see IndividualAPA102Control
+  
+  const uint8_t latch_bytes = 3 * APA102_SPI_BYTES_PER_PIXEL;
+  const unsigned int first_slot = m_start_address - 1;  // 0 offset
+  
+  // check if enough data is there.
+  if (buffer.Size() - first_slot < APA102_SLOTS_PER_PIXEL) {
+    OLA_INFO << "Insufficient DMX data, required " << APA102_SLOTS_PER_PIXEL
+             << ", got " << buffer.Size() - first_slot;
+    return;
+  }
+  
+  // get data for entire string length
+  const unsigned int output_length = m_pixel_count * APA102_SPI_BYTES_PER_PIXEL;
```

Comment:
This should be 

const unsigned int output_length = (1 + m_pixel_count) * APA102_SPI_BYTES_PER_PIXEL;